### ISSUE TITLE
Improve Google Ads error messages

### ISSUE THREAD

**Issue by nataliekwong** [airbytehq/airbyte#26105]
The error messages for Google Ads can be improved to give better guidance to the user setting up the connector.

Type of error: No associated Google Ads accounts associated with user

> Existing check error message: Unable to connect to Google Ads API with the provided configuration - User in the cookie is not a valid Ads user.

Suggestion: The authenticated user is not associated with any Google Ads account. Try authenticating with a different email.

Type of error: Customer ID is entered incorrectly

> Existing toast error message: The value must match the pattern ^[0-9]{10}(,[0-9]{10})*$

Suggestion: "The customer ID must be 10 digits. Separate multiple customer IDs using commas.

> Existing error message: Unable to connect to Google Ads API with the provided configuration - Invalid customer ID '0001917534'.

Suggestion: Failed to access the customer "xxxxxx". Verify the entered customer ID.

> Existing check error message: Unable to connect to Google Ads API with the provided configuration - The customer account can't be accessed because it is not yet enabled or has been deactivated.
> Unable to connect to Google Ads API with the provided configuration - The customer can't be used because it isn't enabled.

Suggestion: Failed to access the customer "xxx".  The customer account is not yet enabled or has been deactivated.

Type of error: Start Date incorrect format

> Existing error message: The value must match the pattern YYYY-MM-DD

No suggestions here

Type of error: Custom GAQL Query incorrect (several examples listed of what the error message appears as)

> Existing error message: incorrect GAQL query statement: '"SELECT campaign.id, campaign.name FROM campaign"'

> Unable to connect to Google Ads API with the provided configuration - Cannot select or filter on the following segments: 'segments.date'(could not support requested resources: 'AD_GROUP_CRITERION'), since segment is incompatible with the resource in the FROM clause or other selected segmenting resources.

> Custom query should not contain segments.date'

>Metrics cannot be requested for a manager account. To retrieve metrics, issue separate requests against each client account under the manager account. 

Suggestion:  The custom GAQL query '{Destination Table Name}' failed. Validate your GAQL query with the Google Ads [query validator](https://developers.google.com/google-ads/api/fields/v13/query_validator). 

Type of error: Login Customer ID for Managed Accounts - customer ID incorrect

> Existing error message: Unable to connect to Google Ads API with the provided configuration - Invalid customer ID '0001917534'.

Suggestion: Failed to access the customer "xxxxxxxxxx". Ensure the customer is linked to your manager account or check your permissions to access this customer account.

Type of error: Conversion Window not correct

> Existing toast message: connectionConfiguration.conversion_window_days must be less than or equal to 1095

Suggestion: The conversion window must be less than or equal to 1095

Type of error: End Date incorrect

> Existing error message: The value must match the pattern YYYY-MM-DD

No suggestions here. 

Type of error: OAuth token expired or invalid:

> Existing error message: Jwt is expired

> RefreshError('invalid_grant: Token has been expired or revoked.', {'error': 'invalid_grant', 'error_description': 'Token has been expired or revoked.'})

Suggestion: The authentication to Google Ads has expired. Re-authenticate to restore access to Google Ads. 

Type of error: Account is deleted
> RefreshError('invalid_grant: Account has been deleted', {'error': 'invalid_grant', 'error_description': 'Account has been deleted'})

Suggestion: The customer account "xxxxxx" has been deleted. Verify the entered customer ID.

Type of error: Stream not found
> The requested stream teste_fato_regiao was not found in the source. Available streams: 

Suggestion: The stream {name} no longer exists in the configuration. Refresh the schema in replication settings and remove this stream from future sync attempts.

**Comment by nataliekwong** [comment]
Some of these are emitted on the connector setup form during the connector check, whereas others are emitted in the connection logs as a source error. If needed, I can divide the errors observed into two issues if they should be handled separately.

**Comment by artem1205** [comment]
Following messages were not added in [this PR](https://github.com/airbytehq/airbyte/pull/26948/files) because I hand not enough context about these errors. We use python GoogleAdsClient , and I need to be sure which exception is raised. I occasionally look at the errors, I’ll fix it as soon as such errors appear:

1. displaying The authenticated user is not associated with any Google Ads account. Try authenticating with a different email. instead of Unable to connect to Google Ads API with the provided configuration - User in the cookie is not a valid Ads user.
2. Failed to access the customer "xxxxxxxxxx". Ensure the customer is linked to your manager account or check your permissions to access this customer account. instead of Unable to connect to Google Ads API with the provided configuration - Invalid customer ID '0001917534'.
3. The customer account "xxxxxx" has been deleted. Verify the entered customer ID. instead of RefreshError('invalid_grant: Account has been deleted', {'error': 'invalid_grant', 'error_description': 'Account has been deleted'})

This one is related to Airbyte CDK, and should be handled separately:
 - The stream {name} no longer exists in the configuration. Refresh the schema in replication settings and remove this stream from future sync attempts. instead of  The requested stream teste_fato_regiao was not found in the source. Available streams:

**Comment by artem1205** [comment]
@nataliekwong,
I suppose we can close this issue for now.
Should we create new separate issues to address changes in CDK and currently missing Google Ads Exceptions?

WDYT?

cc @lazebnyi @oustynova

**Comment by nataliekwong** [comment]
@artem1205  I checked the Google Ads error docs and am curious if these are what you are looking for:
- 1: https://developers.google.com/google-ads/api/reference/rpc/v14/AuthorizationErrorEnum.AuthorizationError
- 2: https://developers.google.com/google-ads/api/reference/rpc/v14/AuthorizationErrorEnum.AuthorizationError
- 3: https://developers.google.com/google-ads/api/reference/rpc/v14/AuthenticationErrorEnum.AuthenticationError

Do those help with identifying the error?

I agree that the CDK change would be a separate issue, can you create and link that here so I can follow it?

**Comment by artem1205** [comment]
Thanks for searching, unfortunately we use [Google Ads Python library](https://github.com/googleads/google-ads-python) in our connector, which means that all our requests (and possible errors) are "wrapped" in Python Exceptions.

I've created new issue for CDK enhancement: https://github.com/airbytehq/airbyte/issues/27393.

**Comment by nataliekwong** [comment]
@artem1205 Thanks for the explanation. Is the issue that you cannot reproduce the situation in which those errors are generated? I was able to generate scenario 2 by authenticating with OAuth and then entering a customer ID I did not have access to. Could you help me understand the issue in addressing these scenarios that happen quite commonly? 

Could you also remove the regex on the pattern descriptor for the Customer ID field? It should be:  "pattern_descriptor": "The customer ID must be 10 digits. Separate multiple customer IDs using commas.", 

Instead of: "pattern_descriptor": "^[0-9]{10}(,[0-9]{10})*$ . The customer ID must be 10 digits. Separate multiple customer IDs using commas.",

**Comment by artem1205** [comment]
@nataliekwong ,

1. Ok, i'll try to reproduce the error using the provided scenario handle the exception. Will get back once finished.  
2. I've already updated `pattern_descriptor` in [PR](https://github.com/airbytehq/airbyte/pull/26948/files#diff-d0c05056417b827c55593343d2cd91f0494702451c542da15477df2876b55e3dR63).

**Comment by nataliekwong** [comment]
The pattern descriptor shouldn't have regex in it (it's not commonly understood by users, so we should use it in behind the form validation, but not presented to users in the error messages themselves.)

**Comment by artem1205** [comment]
@nataliekwong,
if I enter wrong customer ID i get 
``` json
"message": "'Unable to connect to Google Ads API with the provided configuration - No customer found for the provided customer id.'"
```
Which error message should be shown instead?

**Comment by nataliekwong** [comment]
@artem1205 I see the PR is merged - let me know if I should provide more clarity here

**Comment by artem1205** [comment]
@nataliekwong ,
Can we close this issue ?

**Comment by nataliekwong** [comment]
Hey @artem1205 , even though the PR https://github.com/airbytehq/airbyte/pull/27608 is merged, I am still seeing the existing error here:
  
<img width="703" alt="Screen Shot 2023-07-03 at 6 13 53 AM" src="https://github.com/airbytehq/airbyte/assets/38087517/2ffcabdc-d384-49ab-a9e4-c005a4aee133">

What I did: 
I have a successfully authenticated Google Ads source. I changed the customer iD by one digit to a customer ID I don't have access to. I got this error message. I would have expected to see the message: "Failed to access the customer '{exception.customer_id}'. Ensure the customer is linked to your manager account or check your permissions to access this customer account."

I would keep this open if the intended error message has not been implemented.
